### ISSUE TITLE
Is this package abandoned?

### ISSUE THREAD

**Issue by willemstuursma** [thephpleague/flysystem-azure#33]
Not a lot of activity happening here and dependencies on very old other packages. Are you still supporting this @frankdejonge?

**Comment by frankdejonge** [comment]
Hi @willemstuursma I've not abandoned this package, but it does need updating.

**Comment by willemstuursma** [comment]
Thanks @frankdejonge! Anything I can do to help? I have a full Azure set up available, so I could review some of the PR's maybe?

**Comment by maetthu** [comment]
@willemstuursma @frankdejonge Same here, could help as well. Microsoft also just recently abandoned microsoft/azure-storage in favor of dedicated ones per storage product ([microsoft/azure-storage-blob](https://packagist.org/packages/microsoft/azure-storage-file) and [microsoft/azure-storage-file](https://packagist.org/packages/microsoft/azure-storage-file) for file storage), so it might be better to switch to these. Or maybe even a dedicated adapter for file- and blob storage, _AzureBlob_ and _AzureFile_ as these products are pretty different (though I didn't check out the SDK yet, maybe the interfaces are similar enough for a single Azure adapter).

**Comment by frankdejonge** [comment]
It seems those two are read-only splits from this repo: https://github.com/azure/azure-storage-php

**Comment by frankdejonge** [comment]
I've yet to discover what's different about the blob storage and file storage tho.

**Comment by frankdejonge** [comment]
Ok so it seems like blob storage is more like S3 while the file storage is more traditional.

**Comment by maetthu** [comment]
yes, it's only a read-only split, though they abandoned the package from the main repo in [packagist](https://packagist.org/packages/microsoft/azure-storage):
> This package is abandoned and no longer maintained. The author suggests using the microsoft/azure-storage-blob, microsoft/azure-storage-table, microsoft/azure-storage-queue, microsoft/azure-storage-file package instead.

 Blob storage is an object storage and is more or less equivalent to AWS S3, while file storage is a shared filesystem service, similar as AWS EFS, though not NFS but CIFS. It also has a REST API to access files.

**Comment by frankdejonge** [comment]
Ah, so it's advised to directly couple on the splits. Which makes sense.

**Comment by maetthu** [comment]
This [cheatsheet](https://docs.microsoft.com/en-us/azure/architecture/aws-professional/services) might help if you are familiar with AWS terminology.

**Comment by cloelke** [comment]
Looks like  ijin82 has already started to update the dependencies, but it would be nice if blobs or files could be used since Microsoft has added the option to extend local storage to azure-files with a two way sync
https://github.com/ijin82/flysystem-azure/commit/4e0c71f9491a3d2397236b7159cbfb69464556e0

**Comment by frankdejonge** [comment]
I think it'd be better to start a fresh repo for this. I'm in contact with somebody at Microsoft to see if I can get some resources for integration tests in which case I can just build the thing from scratch. Because of the underlying adapter switch and general differences I prefer to have this "update" be a new package.

**Comment by willemstuursma** [comment]
Didn't Microsoft have this PHP community guy, I think I saw hime once on the DPC or PhpBenelux...

**Comment by frankdejonge** [comment]
That might be Josh Holmes actually, which is my contact there.

**Comment by jayhaase** [comment]
Hey Friends! I am working on a quick turn project, and would love to use a modernized version of this adapter. What can I do to help you get a new version released?

**Comment by frankdejonge** [comment]
I've had a Skype call with a representative of Microsoft a couple days ago and they're looking into sponsoring resources. When that's available we can instantly breath new live into this adapter. Otherwise if somebody has an credentials for me I can already start working on this.

**Comment by jayhaase** [comment]
@frankdejonge Does this help at all? https://docs.microsoft.com/en-us/azure/storage/common/storage-use-emulator

I am going to play with https://hub.docker.com/r/arafato/azurite/ and see if I can use it for some testing.

**Comment by frankdejonge** [comment]
@jayhaase that's a windows program, not something I can use unfortunately :(

**Comment by maetthu** [comment]
@frankdejonge I could provide you with an Azure storage account for testing, no problem. Just tell me where I can send it to (and preferably a GPG key as well).

**Comment by jayhaase** [comment]
@frankdejonge as a backup to creds from @maetthu , you could use the docker container -- does not require Windows. I am playing with it now.

**Comment by frankdejonge** [comment]
@jayhaase is that docker container available somewhere?

**Comment by jayhaase** [comment]
@frankdejonge Here is the process to setup an Azure Blob emulator:

1. Install Docker > https://www.docker.com/community-edition
2. The Docker container info is here > https://github.com/arafato/azurite
3. Run this command to download the container: 

```
docker pull arafato/azurite
```

4. Create a directory that will be your local volume for the conatiner.
5. Run this command to start the container, it will create a container attached to your terminal session, with log messages displayed: 

```
docker run -e executable=blob -i -t -p 10000:10000 -v /Users/jayhaase/Desktop/vol:/opt/azurite/folder arafato/azurite
```

4. Run this curl command in another terminal session to create a container (notice log message in your docker container terminal session):
    
```
curl \
  -X PUT \
  -H "$x_ms_date_h" \
  -H "$x_ms_version_h" \
  -H "Content-Length: 0" \
  -H "$authorization_header" \
  "http://0.0.0.0:10000/devstoreaccount1/mycontainer?restype=container"
```

You can connect to the Azure blob emulator with this connection string:

```
DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;BlobEndpoint=http://0.0.0.0:10000;
```

**Comment by jayhaase** [comment]
Some additional useful Docker commands:

**Comment by jayhaase** [comment]
@frankdejonge How is progress? Is there any way I can help?

**Comment by judgej** [comment]
I'm looking for Azure File Service support too, quite urgently. If there is anything I can do to help get this moved along, let me know. I'm using [Azure queues](https://github.com/squigg/azure-queue-laravel) on Laravel, and that package has just moved to the newer `microsoft/azure-storage-queue` package in the last few days, after the older `microsoft/azure-storage` library was abandoned. From what I understand, it was a smooth changeover. I guess it's `microsoft/azure-storage-file` that is needed here.

**Comment by frankdejonge** [comment]
I'm going to see if I can tackle this tomorrow with the help of that docker container. If so this should be pretty straight forward. If not we have to think of something else.

**Comment by frankdejonge** [comment]
@judgej if you need it urgent, you can always implement it yourself so you're not blocked. Since you've got the actual resources to talk against it shouldn't take more than a couple hours to get it done and fully tested for your use-case.

**Comment by maetthu** [comment]
Another way for a quick and dirty solution until the adapter is ready is to run [minio as a gateway to Azure Blob Storage](https://www.minio.io/azure.html), and use the S3 adapter to access it. Works well enough for my use case and is really simple to run (basically just a docker container), although there are [some limitations](https://docs.minio.io/docs/minio-gateway-for-azure), so YMMV.

**Comment by judgej** [comment]
Yeah, I was mentally blocking some time out for the morning to do this, on the cycle home. I just had to ask first so I wouldn't be duplicating any effort.

**Comment by judgej** [comment]
Someone here has taken this driver, and updated it to the newer `microsoft/azure-storage-blob` library:

https://github.com/ijin82/flysystem-azure

They have then added a Laravel service provider, which can be ignored if you don't need that, but it essentially seems to work, though it is in the author's own namespace. I am actually looking to use the `microsoft/azure-storage-file` library rather than blobs, so I have some work to do there.

**Comment by frankdejonge** [comment]
I've started working on this, hope to get it done in an hour or 2. The adapter will be created here, a new one because the main dependencies have altered: https://github.com/thephpleague/flysystem-azure-blob-storage

**Comment by frankdejonge** [comment]
So far the docker image seems to work btw, so that's good news.

**Comment by frankdejonge** [comment]
The docker image is a little quirky around listing blobs, but I'm able to work around that.

**Comment by frankdejonge** [comment]
Progress is here https://github.com/thephpleague/flysystem-azure-blob-storage
It's not done and help is welcome. In particular the copy operation seems to be busted for some reason and the CI won't create the initial account (yay for "works on my docker engine").

**Comment by maetthu** [comment]
@frankdejonge works great when using it with an actual Azure Storage account & unit tests pass. Except a small bug when not using path prefixes, but not sure if this is related to the docker container. See pull request.

**Comment by judgej** [comment]
If it's any use, I have a dev Azure File Storage driver here:

https://github.com/academe/flysystem-azure-file

It's still rough around the edges, without a release or tests, but it works enough for me to start using it next week and debug and tidy up as I go. Being file storage rather than blob storage, it supports directories. It's quite a split from this blob storage because absolutely everything about the API is different, so it's a choice depending on your use-case rather than a replacement. Hmm, I might just drop the word "-storage" on the end just for consistency.

**Comment by jayhaase** [comment]
@frankdejonge What kind of help do you need?
